File: cmd/cmd.js
#!/usr/bin/env node

function loadModule() {
  const path = require("path");
  let func_name = process.argv[1];
  if (func_name.endsWith("cmd.js")) {
    func_name = process.argv[2]
  } else {
    func_name = path.basename(func_name)
  }
  if (!func_name.endsWith(".js")) {
    func_name += '.js'
  }
  const func_path = path.resolve(__dirname, `../js/${func_name}`)
  try {
    return require(func_path);
  } catch {
    console.log(`Invalid js function: ${func_name}`)
    process.exit(1)
  }
}

if (process.env["LLM_FUNCTION_DECLARATE"]) {
  const { declarate } = loadModule();
  console.log(JSON.stringify(declarate(), null, 2))
} else {
  let data = null;
  try {
    data = JSON.parse(process.env["LLM_FUNCTION_DATA"])
  } catch {
    console.log("Invalid LLM_FUNCTION_DATA")
    process.exit(1)
  }
  const { execute } = loadModule();
  execute(data)
}

File: cmd/cmd.rb
#!/usr/bin/env ruby

require 'json'
require 'pathname'

def load_module
  if __FILE__.end_with?("cmd.rb")
    func_name = ARGV[0]
  else
    func_name = Pathname.new(__FILE__).basename.to_s
  end

  func_name += '.rb' unless func_name.end_with?('.rb')
  func_path = File.expand_path("../rb/#{func_name}", __dir__)

  begin
    return require_relative func_path
  rescue LoadError
    puts "Invalid ruby function: #{func_name}"
    exit 1
  end
end

if ENV["LLM_FUNCTION_ACTION"] == "declarate"
  declarate = load_module.method(:declarate)
  puts JSON.pretty_generate(declarate.call)
else
  begin
    data = JSON.parse(ENV["LLM_FUNCTION_DATA"])
  rescue JSON::ParserError
    puts "Invalid LLM_FUNCTION_DATA"
    exit 1
  end
  execute = load_module.method(:execute)
  execute.call(data)
end